// @taiga-ui/addon-commerce/fesm2022/taiga-ui-addon-commerce-components-input-expire.mjs
import * as i0 from '@angular/core';
import { inject, Directive, Input } from '@angular/core';
import * as i1 from '@maskito/angular';
import { MaskitoDirective } from '@maskito/angular';
import { TUI_INPUT_CARD_OPTIONS } from '@taiga-ui/addon-commerce/components/input-card';
import { TUI_MASK_EXPIRE } from '@taiga-ui/addon-commerce/constants';
import * as i2 from '@taiga-ui/core/components/textfield';
import { TuiWithTextfield } from '@taiga-ui/core/components/textfield';
import { tuiMaskito } from '@taiga-ui/kit/utils';

class TuiInputExpire {
    constructor() {
        this.mask = tuiMaskito(TUI_MASK_EXPIRE);
        this.autocomplete = inject(TUI_INPUT_CARD_OPTIONS).autocomplete;
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: TuiInputExpire, deps: [], target: i0.ɵɵFactoryTarget.Directive }); }
    static { this.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "14.0.0", version: "16.2.12", type: TuiInputExpire, isStandalone: true, selector: "input[tuiInputExpire]", inputs: { autocomplete: "autocomplete" }, host: { attributes: { "inputmode": "numeric", "placeholder": "00/00", "translate": "no", "maxlength": "5" }, properties: { "attr.name": "autocomplete ? \"ccexpiryyear\" : null", "autocomplete": "autocomplete ? \"cc-exp\" : \"off\"" } }, hostDirectives: [{ directive: i1.MaskitoDirective }, { directive: i2.TuiWithTextfield }], ngImport: i0 }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: TuiInputExpire, decorators: [{
            type: Directive,
            args: [{
                    standalone: true,
                    selector: 'input[tuiInputExpire]',
                    hostDirectives: [MaskitoDirective, TuiWithTextfield],
                    host: {
                        inputmode: 'numeric',
                        placeholder: '00/00',
                        translate: 'no',
                        maxlength: '5',
                        '[attr.name]': 'autocomplete ? "ccexpiryyear" : null',
                        '[autocomplete]': 'autocomplete ? "cc-exp" : "off"',
                    },
                }]
        }], propDecorators: { autocomplete: [{
                type: Input
            }] } });

/**
 * Generated bundle index. Do not edit.
 */

export { TuiInputExpire };
