// @taiga-ui/kit/esm2022/components/carousel/carousel.directive.mjs
import { isPlatformServer } from '@angular/common';
import { Directive, inject, Input, PLATFORM_ID } from '@angular/core';
import { WA_PAGE_VISIBILITY } from '@ng-web-apis/common';
import { TUI_FALSE_HANDLER, TUI_TRUE_HANDLER } from '@taiga-ui/cdk/constants';
import { tuiIfMap, tuiTypedFromEvent } from '@taiga-ui/cdk/observables';
import { tuiInjectElement } from '@taiga-ui/cdk/utils/dom';
import { BehaviorSubject, combineLatest, EMPTY, interval, map, merge, Observable, } from 'rxjs';
import * as i0 from "@angular/core";
class TuiCarouselDirective extends Observable {
    constructor() {
        super((subscriber) => this.output$.subscribe(subscriber));
        this.el = tuiInjectElement();
        this.platform = inject(PLATFORM_ID);
        this.visible$ = inject(WA_PAGE_VISIBILITY);
        this.duration$ = new BehaviorSubject(0);
        this.running$ = merge(tuiTypedFromEvent(this.el, 'mouseenter').pipe(map(TUI_FALSE_HANDLER)), tuiTypedFromEvent(this.el, 'touchstart').pipe(map(TUI_FALSE_HANDLER)), tuiTypedFromEvent(this.el, 'touchend').pipe(map(TUI_TRUE_HANDLER)), tuiTypedFromEvent(this.el, 'mouseleave').pipe(map(TUI_TRUE_HANDLER)), this.visible$);
        this.output$ = isPlatformServer(this.platform)
            ? EMPTY
            : combineLatest([this.duration$, this.running$]).pipe(tuiIfMap(([duration]) => interval(duration), (values) => values.every(Boolean)));
    }
    set duration(duration) {
        this.duration$.next(Number.isNaN(duration) ? this.duration$.value : duration);
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: TuiCarouselDirective, deps: [], target: i0.ɵɵFactoryTarget.Directive }); }
    static { this.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "14.0.0", version: "16.2.12", type: TuiCarouselDirective, isStandalone: true, inputs: { duration: "duration" }, usesInheritance: true, ngImport: i0 }); }
}
export { TuiCarouselDirective };
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: TuiCarouselDirective, decorators: [{
            type: Directive,
            args: [{
                    standalone: true,
                }]
        }], ctorParameters: function () { return []; }, propDecorators: { duration: [{
                type: Input
            }] } });
